subsys/net/lib/aws_fota/src/aws_fota.c

```diff
@@ -0,0 +1,261 @@
+		update_device_shadow_version(client);
+		/* This expects that the application's mqtt handler will handle
+		 * any situations where you could not connect to the MQTT
+		 * broker.
+		 */
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		aws_jobs_unsubscribe_expected_topics(client, true);
+		break;
+
+	case MQTT_EVT_PUBLISH: {
+		const struct mqtt_publish_param *p = &evt->param.publish;
+		aws_fota_process(topic);
+
+	} break;
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			printk("MQTT PUBACK error %d\n", evt->result);
+			break;
+		}
+		/* check evt->param.puback.message_id */
+		/* We expect that the client
+		break;
+
+	case MQTT_EVT_SUBACK:
+		if (evt->result != 0) {
+			break;
+		}
+
+		/* check evt->param.suback.message_id against SUB_BASE */
+		break;
+
+	default:
+		/* Handling for default case? */
+		break;
+	}
+}
+
+#define ERR_CHECK(err)\
+	do {\
+		if (err) {\
+			return err;\
+		}\
+	} while(0)
+
+int aws_fota_on_publish_evt(struct mqtt_client *const client,
+		     const u8_t * topic,
+		     u32_t topic_len
+		     const u8_t * json_payload)
+{
+	int err;
+	char job_id[JOB_ID_MAX_LEN];
+	if (!strncmp(notify_next_topic,
+		    topic,
+		    MIN(NOTIFY_NEXT_TOPIC_MAX_LEN, topic_len)
+		    )
+	    ) {
+		/* Unsubscribe from notify_next_topic to not recive more jobs
+		 * while processing the current job.
+		 */
+		err = parse_job_document(json_payload,
+					 job_id,
+					 hostname,
+					 file_path);
+		ERR_CHECK(err);
+		err = aws_jobs_unsubscribe_expected_topics(client, true);
+		ERR_CHECK(err);
+		err = aws_jobs_subscribe_job_id_get_topic(job_id);
+		ERR_CHECK(err);
+		err = aws_jobs_subscribe_job_id_update(job_id);
+
+		/* Expect */
+
+	} else if (
+
+}
+
+int construct_notify_next_topic(struct mqtt_client * client, u8_t * topic_buf)
+{
+	int ret = snprintf(topic_buf,
+			   NOTIFY_NEXT_TOPIC_MAX_LEN,
+			   NOTIFY_NEXT_TOPIC_TEMPLATE,
+			   client_id);
+	if (ret >= size) {
+		LOG_ERR("Unable to fit formated string into to allocate "
+			"memory for notify_next_topic");
+		return -ENOMEM;
+	} else if (ret < 0) {
+		LOG_ERR("Formatting error for notify_next_topic");
+			"%d", entry, ret);
+		return ret;
+	}
+	return 0;
+}
+
+
+/**@brief Initialize the AWS Firmware Over the Air library.
+ *
+ * @param app_version Current version number of the application as a \0
+ *	terminated ASCII string.
+ * @param callback Callback for events generated.
+ *
+ * @retval 0 If successfully initialized.
+ * @retval -EINVAL If any of the input values are invalid.
+ *           Otherwise, a negative value is returned.
+ */
+int aws_fota_init(const char *app_version,
+		  struct pollfd *const app_fds,
+		  aws_fota_callback_t cb)
+{
+	if (app_version == NULL || cb == NULL || app_fds == NULL) {
+		return -EINVAL;
+
+	}
+
+	callback = cb;
+	int err = fota_download_init(http_fota_handler);
+	if (err != 0) {
+		LOG_ERR("fota_download_init error %d", err);
+		return err;
+	}
+
+	err = construct_notify_next_topic(
```

Comment:
I guess it's an early WiP :)

Reply:
You are quite correct, the goal is to have it in a working and decent state by 8:00am tomorrow. Hopefully you will be able to do a 2nd/3rd pass @rlubos , the hopes is for it to go in before tagging. Anyways thanks for dropping by and giving a review regardless :1st_place_medal:

Reply:
A little bit more out of WiP now :)